File: LeetCode/0004.go
package leetcode

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	if len(nums2) < len(nums1) {
		return findMedianSortedArrays(nums2, nums1)
	}

	left, right, tol, mid1, mid2 := 0, len(nums1)+1, len(nums1)+len(nums2)+1, 0, 0
	median1, median2 := 0, 0
	for left < right {
		mid1 = left + (right-left)>>1
		mid2 = tol>>1 - mid1

		min1 := math.MinInt32
		if mid1 != 0 {
			min1 = nums1[mid1-1]
		}

		max1 := math.MaxInt32
		if mid1 != len(nums1) {
			max1 = nums1[mid1]
		}

		min2 := math.MinInt32
		if mid2 != 0 {
			min2 = nums2[mid2-1]
		}

		max2 := math.MaxInt32
		if mid2 != len(nums2) {
			max2 = nums2[mid2]
		}

		if min1 <= max2 {
			median1 = max(min1, min2)
			median2 = min(max1, max2)
			left = mid1 + 1
		} else {
			right = mid1
		}
	}
	if (len(nums1)+len(nums2))%2 == 0 {
		return float64(median1+median2) / 2.0
	}
	return float64(median1)
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

File: leetcode/0004.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int m = nums1.size(), n = nums2.size();
        vector<int> s1(nums1), s2(nums2);
        if (m > n) {
            swap(m, n);
            swap(s1, s2);
        }

        int l = 0, r = m;
        while (l <= r) {
            int i = l + (r - l) / 2;
            int j = (m + n + 1) / 2 - i;
            if (i < m && s1[i] < s2[j - 1])
                l = i + 1;
            else if (i > 0 && s1[i - 1] > s2[j])
                r = i - 1;
            else {
                cout << i << ' ' << j << endl;
                double ans = 0.0;
                if (i == 0)
                    ans = s2[j - 1];
                else if (j == 0)
                    ans = s1[i - 1];
                else
                    ans = max(s1[i - 1], s2[j - 1]);
                cout << ans << endl;
                if ((m + n) & 1) return ans;
                if (i == m)
                    ans += s2[j];
                else if (j == n)
                    ans += s1[i];
                else
                    ans += min(s1[i], s2[j]);
                return ans / 2.0;
            }
        }
        return 0.0;
    }
};